Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp

```diff
@@ -126,6 +127,51 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
     return new_array_buffer_object;
 }
 
+// 1.3.5 ArrayBuffer.prototype.resize ( newLength ), https://tc39.es/proposal-resizablearraybuffer/#sec-arraybuffer.prototype.resize
+JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::resize)
+{
+    // 1. Let O be the this value.
+    auto* array_buffer_object = TRY(typed_this_value(global_object));
+
+    // 2. Perform ? RequireInternalSlot(O, [[ArrayBufferMaxByteLength]]).
+    if (!array_buffer_object->max_byte_length().has_value()) {
+        // 10.1.15 RequireInternalSlot ( O, internalSlot ), https://tc39.es/ecma262/#sec-requireinternalslot
+        // 1. If Type(O) is not Object, throw a TypeError exception.
+        // 2. If O does not have an internalSlot internal slot, throw a TypeError exception.
+        return vm.throw_completion<TypeError>(global_object, ErrorType::NotAResizableArrayBuffer);
+    }
+
+    // 3. If IsSharedArrayBuffer(O) is true, throw a TypeError exception.
+    // FIXME: Check for shared buffer
+
+    // 4. If IsDetachedBuffer(O) is true, throw a TypeError exception.
+    if (array_buffer_object->is_detached())
+        return vm.throw_completion<TypeError>(global_object, ErrorType::DetachedArrayBuffer);
+
+    // 5. Let newByteLength be ? ToIntegerOrInfinity(newLength).
+    auto new_byte_length = TRY(vm.argument(0).to_integer_or_infinity(global_object));
+
+    // 6. If newByteLength < 0 or newByteLength > O.[[ArrayBufferMaxByteLength]], throw a RangeError exception.
+    if (new_byte_length < 0 || new_byte_length > array_buffer_object->max_byte_length().value())
+        return vm.throw_completion<RangeError>(global_object, ErrorType::NewByteLengthOutOfRange);
+
+    // 7. Let hostHandled be ? HostResizeArrayBuffer(O, newByteLength).
+    auto host_handled = TRY(array_buffer_object->host_resize_array_buffer(global_object, new_byte_length));
+
+    // 8. If hostHandled is handled, return undefined.
+    VERIFY(host_handled);
+    return js_undefined();
+
+    // 9. Let oldBlock be O.[[ArrayBufferData]].
+    // 10. Let newBlock be ? CreateByteDataBlock(newByteLength).
+    // 11. Let copyLength be min(newByteLength, O.[[ArrayBufferByteLength]]).
+    // 12. Perform CopyDataBlockBytes(newBlock, 0, oldBlock, 0, copyLength).
+    // 13. NOTE: Neither creation of the new Data Block nor copying from the old Data Block are observable. Implementations reserve the right to implement this method as in-place growth or shrinkage.
+    // 14. Set O.[[ArrayBufferData]] to newBlock.
+    // 15. Set O.[[ArrayBufferByteLength]] to newLength.
```

Comment:
Do we need to implement this step, or is it handled elsewhere?

Reply:
Originally I thought host provided was just spec talk for it being implementation defined and this was some fallback version. Now that I understand what host provided the actual resize logic has been moved here